ENH: sparse: add nD support for element-wise and reduction operations

anushkasuyal opened the pull request:
_Context: This is the third PR (follow-up to https://github.com/scipy/scipy/pull/21197 and #21435) aiming to extend support for n-D COO arrays in SciPy (refer to https://github.com/scipy/scipy/issues/21169). I am undertaking this work as part of my summer internship with Quansight Labs till the end of September, under the mentorship of @rgommers and @izaid._

This PR continues the extension of COO sparse arrays to n-dimensions, by implementing support for element-wise operations such as boolean comparators, `multiply`, `maximum`, `minimum` etc. between n-D COO arrays. Previously, these operations were possible for arrays of the same shape, but now, as we move towards aligning the sparse array behaviour with that of the array API standard, we allow for broadcasting (which was introduced in the PR #21435) of arrays of different shapes so that the above operations can be performed on them.

After CSR `_broadcast_to` is added in #21412, we can now broadcast CSR arrays too. Since this PR implements broadcasting for COO arrays, it would introduce inconsistencies if COO arrays differed in behaviour from other formats. Therefore, this PR also adds broadcasting to these operations for the CSR format.

To keep any addition of compiled code to a minimum, the basic approach followed by most of the methods implemented in this PR is to utilise NumPy's `ravel`/`unravel` and `reshape` tools to map n-D arrays to 2-D and use the pre-existing CSR sparsetools. Since COO allows for quick conversion to CSR, and CSR allows for faster arithmetic operations, this approach would prevent introducing more C++ functions while also utilising the CSR methods that have already been implemented and rigorously tested in the codebase.

We also implement reduction operations such as `max`, `min`, `argmax`, `argmin`, `nanmax`, `nanmin`, `sum`, and `mean` in this PR. Although these are not element-wise operations, they use an approach similar to that of element-wise operations - where we map to 2-D, route via CSR, and map back to n-D.

(Note- diff may be easier to view once #21435 and #21412 have been merged to main.)

anushkasuyal committed:
WIP: n-D dot via 2-D CSR

anushkasuyal committed:
WIP: n-D dot via 2-D CSR

anushkasuyal committed:
WIP: n-D dot via 2-D CSR

anushkasuyal committed:
WIP: n-D dot via 2-D CSR

anushkasuyal committed:
WIP: coord permutation issues fixed

anushkasuyal committed:
WIP: coord permutation issues fixed

anushkasuyal committed:
WIP: coord permutation issues fixed

anushkasuyal committed:
WIP: coord permutation issues fixed

anushkasuyal committed:
WIP: tensordot working(?)

anushkasuyal committed:
WIP: tensordot working(?)

anushkasuyal committed:
WIP: tensordot working(?)

anushkasuyal committed:
WIP: tensordot working(?)

anushkasuyal committed:
dot and tensordot working and tests added

anushkasuyal committed:
dot and tensordot working and tests added

anushkasuyal committed:
dot and tensordot working and tests added

anushkasuyal committed:
dot and tensordot working and tests added

anushkasuyal committed:
_block_diag added along with test

anushkasuyal committed:
_block_diag added along with test

anushkasuyal committed:
_block_diag added along with test

anushkasuyal committed:
_block_diag added along with test

anushkasuyal committed:
_matmul_sparse _block_diag and _extract_block_diag added, _matmul_dispatch modified

anushkasuyal committed:
_matmul_sparse _block_diag and _extract_block_diag added, _matmul_dispatch modified

anushkasuyal committed:
_matmul_sparse _block_diag and _extract_block_diag added, _matmul_dispatch modified

anushkasuyal committed:
_matmul_sparse _block_diag and _extract_block_diag added, _matmul_dispatch modified

anushkasuyal committed:
added and parametrized tests for _block_diag _extract_block_diag and _matmul_sparse

anushkasuyal committed:
coo_matmat_dense_nd added in coo.h, _matmul_multivector modified, tests for the same added

anushkasuyal committed:
error check added in _matmul_dispatch

anushkasuyal committed:
error check added in _matmul_dispatch

anushkasuyal committed:
sparse-dense tensordot and dot added

anushkasuyal committed:
sparse-dense tensordot and dot added

anushkasuyal committed:
sparse-dense tensordot and dot added

anushkasuyal committed:
broadcast_to tests added

anushkasuyal committed:
broadcast_to tests added

anushkasuyal committed:
broadcast_to fixed

anushkasuyal committed:
broadcast_to fixed

anushkasuyal committed:
broadcast_to fixed

anushkasuyal committed:
diagonal working, tests added

anushkasuyal committed:
broadcast_to added for CSR matrices

anushkasuyal committed:
correction made in broadcast_to, tests added

anushkasuyal committed:
boolean comparison = != < <= > >= implemented

anushkasuyal committed:
CSR broadcasting added in multiply and bitwise boolean ops, tests for multiply and comparisons added

anushkasuyal committed:
maximum minimum added for COO, CSR code modified for _maximum_minimum to incorporate broadcasting

anushkasuyal committed:
CSR sparse-sparse ops updated to include broadcasting
this has been done for each method individually so far. following commit(s) would remove these additions from each function and instead incorporate them into _binopt.
NOTE - sparse-dense doesn't support broadcasting yet. _base.py needs some inconsistent shapes error checks to be removed, test_base may need removal of shape compatibility checks too, given we are broadcasting now.

anushkasuyal committed:
TYPO: missing return

anushkasuyal committed:
try /except ValueError added for broadcasting
this was done to prevent test_base ValueError checks from having an error msg mismatch. assert_raises ValueError was commented out from test_base's test_size_zero_matrix-arithmetic test, because broadcasting means size zero matrix arithmetic is now possible.

anushkasuyal committed:
broadcasting checks moved to _base.py from _compressed.py

anushkasuyal committed:
broadcasting moved to _binopt, sparse-sparse add sub working w broadcasting

anushkasuyal committed:
commented out ValueError check
not applicable anymore since we are broadcasting now

anushkasuyal committed:
WIP: max min

anushkasuyal committed:
eliminate_zeros removed

anushkasuyal committed:
.copy() added

anushkasuyal committed:
added NotImplementedError for CSC

anushkasuyal committed:
lint error fixed

anushkasuyal committed:
argmax argmin working, tests for both added

anushkasuyal committed:
validateaxis modified, sum mean working without out param for n-D, tests added and removed for min max sum mean

anushkasuyal committed:
sum mean methods in _coo.py

anushkasuyal committed:
sum mean updated

anushkasuyal committed:
divide added (to be moved to separate PR later)

anushkasuyal committed:
mean WIP

anushkasuyal committed:
sum, mean working with out param, tests modified

anushkasuyal committed:
hstack vstack working, tests added

anushkasuyal committed:
block_diag added, tests added for n-D power, abs, round

anushkasuyal committed:
triu tril find extended to n-D in _extract.py

anushkasuyal committed:
broadcast_to moved to _csr_base in _csr.py

anushkasuyal committed:
np.broadcast_shapes used

anushkasuyal committed:
add self.sum_duplicates()

anushkasuyal committed:
adding copy=False

anushkasuyal committed:
copy argument passed in broadcast_to

anushkasuyal committed:
reverted ValueError condition, modified tests

anushkasuyal committed:
lots of refactoring for matmul, tensordot, dot

anushkasuyal committed:
lots of refactoring for matmul, tensordot, dot

anushkasuyal committed:
if condition fixed

anushkasuyal committed:
tests modified

anushkasuyal committed:
linting issues fixed

[lint only]

anushkasuyal committed:
missing explicit arg for 1-D/2-D added back

anushkasuyal committed:
test parametrize edited

anushkasuyal committed:
removed 2D tests from test_coo.py

anushkasuyal committed:
zero dim size test modified

anushkasuyal committed:
A, B in _matmul_sparse renamed self, other

anushkasuyal committed:
broadcast_to case added for when final shape == self.shape

anushkasuyal committed:
A, B renamed to self, other

anushkasuyal committed:
lint error fixed

anushkasuyal committed:
replace vstack

anushkasuyal committed:
avoid copying

anushkasuyal committed:
remove duplicate function

anushkasuyal committed:
broadcast_to made private

anushkasuyal committed:
broadcast_to test edited

anushkasuyal committed:
Revert "replace vstack"

This reverts commit 915e3cb4c1529ff1594bdf85c8aeea32d7dafd65.

anushkasuyal committed:
vstack replaced

anushkasuyal committed:
array conversion avoided

anushkasuyal committed:
rmatmul added

anushkasuyal committed:
add test condition for sparse-dense

anushkasuyal committed:
Update scipy/sparse/sparsetools/coo.h

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
small refactoring changes to _matmul_dispatch

anushkasuyal committed:
Update scipy/sparse/tests/test_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
error refactoring for COO dot

anushkasuyal committed:
copy arg added to _broadcast_to

anushkasuyal committed:
1d reshaping for matmul

anushkasuyal committed:
matmul CI failure fixed

anushkasuyal committed:
C++ array issue fixed

anushkasuyal committed:
vstack replaced with tuple

anushkasuyal committed:
array/tuple refactoring in _sparse_tensordot

anushkasuyal committed:
array/tuple changes in _broadcast_to

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
iter_cs added for unravelling

anushkasuyal committed:
og_shape_self name changed

anushkasuyal committed:
add dim_nnz for efficiency boost

anushkasuyal committed:
post rebase modifications

anushkasuyal committed:
refactoring

anushkasuyal committed:
triu tril removed, moved to PR4

anushkasuyal committed:
lint checks

anushkasuyal commented:
After incorporating the methods extended in this PR, the updated tracker would be as follows -

| **Methods**                              | **1D/2D** | **nD** | **Implementation**                                      |
| ------------------------------------ | ---------- | --- | --------------------------------------------------- |
| `reshape`                              | Y          | Y   | all-D in python                                   |
| `count_nonzero`                        | Y          | Y   | all-D in python                                   |
| `transpose`                            | Y          | Y   | all-D in python                                   |
| `resize`                               | Y          |     | 1- and 2-D in python                               |
| `toarray`                              | Y          | Y   | 2-D uses `coo_todense`; 1- and n-D use `coo_todense_nd` |
| `tocsc`                                | Y          | \-  |                                                     |
| `tocsr`                                | Y          | \-  |                                                     |
| `tocoo`                                | Y          | Y   |                                                     |
| `todia`                                | Y          | \-  |                                                     |
| `todok`                                | Y          | \-  |                                                     |
| `diagonal`                             | Y          |     | in python                                           |
| `sum_duplicates`                       | Y          | Y   | all-D in python                                       |
| `eliminate_zeros`                      | Y          | Y   | all-D in python                                       |
| `_add_dense`                           | Y          | Y   | 1- and 2-D use `coo_todense`; n-D uses `coo_todense_nd` |
| `_add_sparse`                          | Y          | Y   | 1- and 2-D route via CSR; n-D uses python           |
| `_sub_dense`                           | Y          | Y   | all-D convert to dense and subtract                 |
| `_sub_sparse`                          | Y          | Y   | 1- and 2-D route via CSR; n-D uses python           |
| ` _matmul_vector` (sparse-dense)     | Y          | Y   | 1- and 2-D use `coo_matvec`; n-D uses `coo_matvec_nd`   |
| `_matmul_multivector` (sparse-dense) | Y          | Y | 1- and 2-D use `coo_matmat_dense`; n-D uses `coo-matmat-dense-nd` |
| `_matmul_sparse` (sparse-sparse)     | Y          |  Y   | 1- and 2-D route via CSR; n-D maps to 2-D then routes via CSR         |
| `multiply`                             | Y          |  Y   | via CSR                                             |
| `maximum`                              | Y          | Y    | via CSR                                             |
| `minimum`                              | Y          | Y    | via CSR                                             |
| `max`                                  | Y          | Y    | via 2-D implementation  |
| `min`                                  | Y          |  Y   | via 2-D implementation   |
| `nanmax`                               | Y          |  Y   | via 2-D implementation   |
| `nanmin`                               | Y          |  Y   | via 2-D implementation   |
| `power`                                | Y          | Y   | implemented for all-D in `_data_matrix`  |
| `eq`                                   | Y          |  Y   | via CSR                                             |
| `ne`                                   | Y          |  Y   | via CSR                                             |
| `lt`                                   | Y          |  Y   | via CSR                                             |
| `gt`                                   | Y          |  Y   | via CSR                                             |
| `le`                                   | Y          |  Y   | via CSR                                             |
| `ge`                                   | Y          |  Y   | via CSR                                             |
| `abs`                                  | Y          | Y   | uses built-ins in `_data_matrix`                     |
| `round`                                | Y          | Y   | uses np.around in `_data_matrix`                     |
| `div`                                  | Y          |     | uses numpy and routes via CSR                       |
| `dot` (sparse-dense)                   | Y          | Y  | 1- and 2-D uses matmul; map n-D to 2-D and use `csr_matmat` |
| `dot` (sparse-sparse)                 | Y          | Y  | 1- and 2-D uses matmul; reshape then use n-D `matmul` |
| `tensordot` (sparse-dense)             |   Y     |  Y   |   map n-D to 2-D and use `csr_matmat`        |
| `tensordot` (sparse-sparse)            |  Y      |  Y   |   reshape then use n-D `matmul`        |
| `broadcast_to`                    |  Y      |  Y   |   all-D use numpytools                |
| `todense`                          | Y            | Y    |                                                       |
| `random_array`                          | Y            | Y    |                                                       |
| `hstack`                          |  Y           |     |                                                       |
| `vstack`                          |  Y           |     |                                                       |
| `block_array`                          | Y            |     |                                                       |
| `sum`                          |  Y           | Y    |                          via CSR                     |
| `mean`                          | Y            | Y   |                   via CSR                            |
| `triu`                          | Y            |     |                                                       |
| `tril`                          | Y            |     |                                                       |

anushkasuyal committed:
nD diagonal removed, moved to PR 4

dschult commented:
Very nice to see this coming along!!  There are some tests that are failing now, but I suspect they will not take long to track down.

I don't know if this suggestion would make it easier or harder to implement this PR, but the code for `csr._broadcast_to` would be fairly easy to move to `_compressed.py` and make work for CSC as well as CSR. You would need to use `self._swap(shape)` in places where the code accesses `indptr` and `indices` so the csc rows use `indices` and columns use `indptr`.

It's also not clear to me that we want to handle broadcasting by switching formats to csr and back. But it is a feature that results in an exception currently, so implementing it via switching formats is perhaps better than not providing anything. 
But it might be better for the user to convert explicitly to CSR and run the whole expression rather than switching back and forth potentially many times. This PR lets us fairly easily explore that (amazing!) 

It would probably be easier to review if the 2D broadcasting feature was separated from the nD feature, but I know there are time constraints. And all this is building on PRs that haven't merged yet. That will simplify the diff quite a bit -- and will require merging into this one almost certainly. :)  Thanks so much for this. It is a lot of help.

anushkasuyal committed:
int32 error fixed

anushkasuyal committed:
inconsistent shapes error fixed

dschult commented:
Conversions are linear in array size so not too expensive, but we still try to avoid them unless the new format is sufficiently more efficient.

It looks like `_broadcast_to` immediately converts to CSR, broadcasts, and then converts the result back to the original format. I believe that is different from how, e.g. converting to COO for `reshape` or converting to CSR for arithmetic work. They convert format just before the operation and they leave the result in that new format. There is no conversion back.

So an expression containing 5 operations could require 5 conversions (but usually many fewer because the results are usually CSR). With broadcasting as currently in this PR, we have two additional conversions during the broadcast preparation for each operation. So that might create 15 conversions in all (probably fewer but two required for any broadcast of a non-CSR format).

There are two aspects of this:
- `_broadcast_to` converts back to the original format
- `_broadcast_to` immediately converts to CSR before checking if any work is needed. I think this is partly because it is called before the operation code is called. So it is hard to know whether any work is needed.

Proposal:
-  put the broadcast as close to the operation as possible. That is, only the CSR element-wise binary operations should do the broadcasting. I think this is mostly `_binopt`, `_add_dense`, maybe `_binopt_scalar`, and then the three places where `TODO sparse broadcasting` appear: `__eq__`, `__ne__` and `_inequality`. It looks like `COO._add_dense` is the only place outside of CSR, but I might have missed some.
- move the shape checking logic currently in `_base.py` for `__add__` and `__sub__` to the checks inside CSR's `_binopt`, `_add_dense` etc.
- change `_spbase._broadcast_to` so it doesn't convert back to the current format.

I think you have done much of this already in this PR, but many shape checking code spots can be removed if we rely on the broadcasting/shape checking code in `_binopt`. Before jumping in to do this -- check this proposal. Does it make sense to do? Are there corner cases I've missed? Do you think there's a better way to reduce conversions?

Notes: 
- the current `_binopt` calls `to_csr` in lots of places where it is not needed once the `broadcast_to` doesn't convert back.
- we should really move the `_broadcast_to` method to `_compressed.py` so it can work for CSC as well as CSR formats.

anushkasuyal committed:
_broadcast_to unnecessary conversion avoided

anushkasuyal committed:
typo

anushkasuyal committed:
refactoring broadcasting in _binopt etc.

anushkasuyal commented:
Thanks a lot @dschult ! I've read through your proposal and I agree with it. I've also roughly implemented some of it to ensure its validity and it seems to work fine! I've addressed some of your suggestions individually below -

> It looks like `_broadcast_to` immediately converts to CSR, broadcasts, and then converts the result back to the original format. I believe that is different from how, e.g. converting to COO for `reshape` or converting to CSR for arithmetic work. They convert format just before the operation and they leave the result in that new format. There is no conversion back.

Ah yes, true. Let's make `_broadcast_to` not convert back to the original format then. Especially since it's not a user-facing function, it doesn't really matter if we don't convert back.
Also, we can avoid `_broadcast_to` immediately converting to CSR before checking if any work is needed by doing something like this maybe -
```py3
    def _broadcast_to(self, shape, copy=False):
        if self.shape == shape:
            return self.copy() if copy else self
        else:
            return self.tocsr()._broadcast_to(shape, copy)
```
I've made and pushed the above changes.

I've also moved the `np.broadcast_shapes` checks to `_binopt_` and `_add_dense` and removed them from everywhere else (where they were redundant!)


Next, I'll have a think about the swap approach you mentioned earlier for CSC broadcasting.

dschult commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,59 @@ def tobsr(self, blocksize=None, copy=True):
+            data = np.full(N, self.data[0])
+            indices = np.arange(0,N)
+            indptr = np.array([0, N])
+            return self.__class__((data, indices, indptr), shape=shape, copy=False)
```
Make this CSR because 1D not supported by CSC and wont be. So only have to work with CSR
(`self._fmt_container` is either a `spmatrix` or a `sparray` with format "fmt")
```suggestion
            return self._csr_container((data, indices, indptr), shape=shape, copy=False)
```

dschult commented on scipy/sparse/_csr.py:
```diff
@@ -122,6 +122,59 @@ def tobsr(self, blocksize=None, copy=True):
+        if self.nnz == 0: # array has no non zero elements
+            return self.__class__(shape, dtype=self.dtype, copy=False)
+        
+        self.sum_duplicates()
+        if all(s == 1 for s in old_shape):
+            # Broadcast a single element to the entire shape
+            data = np.full(shape[0] * shape[1], self.data[0])
+            indices = np.tile(np.arange(shape[1]), shape[0])
+            indptr = np.arange(0, len(data) + 1, shape[1])
+        
+        elif old_shape[0] == 1 and old_shape[1] == shape[1]:
+            # Broadcast row-wise
+            data = np.tile(self.data, shape[0])
+            indices = np.tile(self.indices, shape[0])
+            indptr = np.arange(0, len(data) + 1, len(self.data))
+
+        elif old_shape[1] == 1 and old_shape[0] == shape[0]:
+            # Broadcast column-wise
+            data = np.repeat(self.data, shape[1])
+            indices = np.tile(np.arange(shape[1]), len(self.data))
+            indptr = self.indptr * shape[1]
+
+        return self.__class__((data, indices, indptr), shape=shape, copy=False)
```
The idea here is to swap shape, and the concept of row, column when workign with indices and indptr.
So the code the handled rows for CSR will be handling columns for CSC and vice versa.
I made the shapes `M, N` and `oM,oN` so I can easily think about code in terms of CSR (M as rows and N as cols) and I let swap handle the CSC case. And I don't have to change the actual `shape` or `old_shape` to make the code work. Leave them as the true shapes to be used at the end when constructing the result. I'm really only changing `shape[...]` to `M` or `N`, and `old_shape[...]` to 'oM` and `oN`. And I adjust the comments for CSC.
```suggestion
        self.sum_duplicates()
        M, N = self._swap(shape)
        oM, oN = self._swap(old_shape)
        if all(s == 1 for s in old_shape):
            # Broadcast a single element to the entire shape
            data = np.full(M * N, self.data[0])
            indices = np.tile(np.arange(N), M)
            indptr = np.arange(0, len(data) + 1, N)

        elif oM == 1 and oN == N:
            # Broadcast row-wise (columns for CSC)
            data = np.tile(self.data, M)
            indices = np.tile(self.indices, M)
            indptr = np.arange(0, len(data) + 1, len(self.data))

        elif oN == 1 and oM == M:
            # Broadcast column-wise (rows for CSC)
            data = np.repeat(self.data, N)
            indices = np.tile(np.arange(N), len(self.data))
            indptr = self.indptr * N

        return self.__class__((data, indices, indptr), shape=shape, copy=False)
```
Of course these changes need to be followed by moving this function to `_compressed.py` so CSC inherits it. AND `_bsr.py` needs to overwrite `_broadcast_to` to use `_spbase._broadcast_to` instead of the version in `_cs_matrix`.

anushkasuyal replied:
Thanks a lot! I've made these changes at https://github.com/scipy/scipy/pull/21613/commits/050a465b7ffc84c9851935249d82b213ed8e2ab8 and I've added tests to `_csc.py` too! Is it okay to have separate tests in `_csc.py` and `_csr.py`? (since having them in `_base.py` caused some int32 issues in one of the dtype tests because we had sparse array objects (along with matrix) in those tests.)

I also cherry-picked this commit to the CSR broadcasting PR, so if it's easier you can review them there, and then once that's merged I'll just rebase this one.

dschult replied:
> Is it okay to have separate tests in _csc.py and _csr.py? (since having them in _base.py caused some int32 issues in one of the dtype tests because we had sparse array objects (along with matrix) in those tests.)
Can you help me understand this? Are you saying that putting the test method into `test_base.py` caused int32 issues? Which test class did you put the method into? I'm not saying you need to put the tests in there. But I'd like to know where there might be int32 issues.  Thanks!

anushkasuyal replied:
It refers to this conversation (https://github.com/scipy/scipy/pull/21412#discussion_r1765053922) we had earlier in the other PR, when some CI tests were failing so we moved the _broadcast_to tests to test_csr.py. I assumed that might've been because 1D CSR used sparse arrays instead of matrices. I can move the tests to test_base.py to see if there are any failures for CSC too.

dschult reviewed (commented):
Here are some suggestions for changing the `CSR` version to work for either CSR or CSC.  I formatted them as suggestions, but you probably shouldn't "commit suggestion" because the whole function needs to be moved to `_compressed.py` to allow CSC to inherit them.  So just make these changes to the function locally and move it to the other module.

AND THEN:  Of course these changes need to be protected in `_bsr.py`. That is, we don't want BSR to inherit these, so we need to overwrite `_broadcast_to` in `bsr.py` to use `_spbase._broadcast_to` instead of the version in `_cs_matrix`.

You could make these changes to #21412 if that is a better PR to do it. And really, these could wait for another PR because they are performance improvements. The current PR has CSR working through `_spbase` which converts to CSR.

Warning: These suggestions are not tested... probably better to change locally rather than committing these via a github option.

anushkasuyal committed:
Update scipy/sparse/_csr.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
CSC broadcasting added, _broadcast_to moved to _compressed.py

anushkasuyal committed:
tests moved

anushkasuyal committed:
lint errors

dschult commented on scipy/sparse/_base.py:
```diff
@@ -1146,6 +1158,17 @@ def sum(self, axis=None, dtype=None, out=None):
         # For some sparse array formats more efficient methods are
         # possible -- these should override this function.
+
+        # if 2-d and axis is tuple
+        if self.ndim == 2 and type(axis) is tuple and len(axis) == 2:
+            # adjust axes
+            axis = [ax if ax>=0 else ax+2 for ax in axis]
```
I think this does what you want without requiring changes later for nD case
```suggestion
        # if axis is tuple of all dimensions then same as axis=None
        if type(axis) is tuple and len(axis) == self.ndim:
            # adjust axes
            axis = [ax if ax >= 0 else ax + self.ndim for ax in axis]
```

dschult commented on scipy/sparse/_base.py:
```diff
@@ -1233,13 +1264,30 @@ def mean(self, axis=None, dtype=None, out=None):
             res = inter_self / self.shape[-1]
             return res.sum(dtype=res_dtype, out=out)
 
+        # if 2-d and axis is tuple
+        if self.ndim == 2 and type(axis) is tuple:
+            # adjust axes
+            axis = [ax if ax>=0 else ax+2 for ax in axis]
+            if len(axis) == 2:
```
Can this be changed similar to above to remove dependence on ndim==2?

anushkasuyal replied:
Done at 8d67717309f9a85f42a7cc182f883c2392f31ef3.

dschult commented on scipy/sparse/_compressed.py:
```diff
@@ -369,7 +386,13 @@ def __ge__(self, other):
     def _add_dense(self, other):
         if other.shape != self.shape:
-            raise ValueError(f'Incompatible shapes ({self.shape} and {other.shape})')
+            try:
+                # This will raise an error if the shapes are not broadcastable
+                bshape = np.broadcast_shapes(self.shape, other.shape)
+            except ValueError:
+                raise ValueError('inconsistent shapes')
```
Can we just use the numpy error messages about broadcasting not being possible? Or are they not helpful in this situation. It would remove quite a few try/excepts leaving a 1 line instead of 5 in each case.

The message for `np.broadcast_shapes` looks like:
`ValueError: shape mismatch: objects cannot be broadcast to a single shape.  Mismatch is between arg 0 with shape (2, 3) and arg 1 with shape (5, 3).`

anushkasuyal replied:
Done at d05853538012f7ab99ebce49d6d51d77c273c821.

dschult commented on scipy/sparse/_compressed.py:
```diff
@@ -384,6 +407,7 @@ def _add_sparse(self, other):
     def _sub_sparse(self, other):
         return self._binopt(other, '_minus_')
 
+
```
extra blank line?

anushkasuyal replied:
Removed

dschult commented on scipy/sparse/_compressed.py:
```diff
@@ -624,8 +648,8 @@ def _maximum_minimum(self, other, npop, op_name, dense_check):
             return npop(self.todense(), other)
-        elif issparse(other):
-            return self._binopt(other, op_name)
+        elif issparse(other):       
+            return self._binopt(other, op_name) 
```
Looks like added spaces at end of line.  That happens elsewhere too.

anushkasuyal replied:
Removed from there and a few other places

dschult commented on scipy/sparse/_data.py:
```diff
@@ -209,8 +209,24 @@ def _min_or_max_axis(self, axis, min_or_max, explicit):
+        if type(axis) is tuple and len(axis) == 1:
+            axis = axis[0]
+
+        if self.ndim == 2 and type(axis) is tuple:
+            if len(axis) == 2:
+                axis = [ax if ax>=0 else ax+2 for ax in axis]
```
```suggestion
        if type(axis) is tuple and self.ndim == len(axis):
                axis = [ax if ax >= 0 else ax + self.ndim for ax in axis]
```

dschult commented on scipy/sparse/tests/test_arithmetic1d.py:
```diff
@@ -307,14 +307,14 @@ def test_size_zero_matrix_arithmetic(self, spcreator):
         mat = np.array([])
         a = mat.reshape(0)
-        d = mat.reshape((1, 0))
+        #d = mat.reshape((1, 0))
```
Why are these tests commented out?

anushkasuyal replied:
Sorry, I meant to remove them altogether, not just comment them out! This one gave a ValueError before for inconsistent shapes but now that we broadcast during addition too, the error wouldn't be raised anymore. Removed at 991fbaa052b749cf888f0b62aa2e4f249144c196.

dschult replied:
Let's leave this test in place (it should not raise an error any more because broadcasting is supported here so it is helpful to test that does not raise an error).
I'm going to try to resolve conflicts here and then make this change. Most of the conflicts were caused by my index_dtype handling changes so I should be able to figure them out.

dschult reviewed (commented):
I'm pushing the time limit for you, but university hasn't started yet... :)
Here are some more comments to look through.
Thanks again for all this work. I will try to continue this effort.

anushkasuyal committed:
axis change generalised

anushkasuyal committed:
Update scipy/sparse/_base.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
extra line and spaces removed

anushkasuyal committed:
Update scipy/sparse/_data.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
test removed

anushkasuyal committed:
default broadcasting to be handled by numpy, "inconsistent shapes" error checks changed to "cannot be broadcast"

dschult commented on scipy/sparse/_compressed.py:
```diff
@@ -1371,17 +1396,16 @@ def _binopt(self, other, op):
-        if other.shape != self.shape:
-            raise ValueError('inconsistent shapes')
-
+        """ 
```
extra space at the end of the line

dschult reviewed (commented):
Some of these broadcasts are not needed. When `self > other` and other is dense, we just convert self to dense. So both are numpy arrays and we can/should rely on numpy to do the broadcasting. 

Two advantages of using numpy broadcasting here:
- we don't have to create the larger array (numpy will use a view instead of a copy).
- numpy will be able to process a smaller array and know what broadcasting shortcuts are possible.

anushkasuyal committed:
sparse-dense CS broadcasting removed

anushkasuyal commented on scipy/sparse/_compressed.py:
```diff
@@ -289,18 +292,20 @@ def __ne__(self, other):
+            bshape = np.broadcast_shapes(self.shape, other.shape)
+            
+            self = self._broadcast_to(bshape)
+            other = np.broadcast_to(other, bshape)
```
Addressed here for all 3 suggestions - 4924acfb36823aea024af231df7b2c56ce2885da

anushkasuyal committed:
extra spaces / lines removed

anushkasuyal commented:
> When `self > other` and other is dense, we just convert self to dense. So both are numpy arrays and we can/should rely on numpy to do the broadcasting.

Just to confirm, do you mean the case when `self.ndim > other.ndim`?

Does that mean that in `_coo_base` we check first if `self.ndim > other.ndim` and if so we perform the operation between `self.todense()` and `other`, otherwise we carry on as it is?

dschult commented:
Sorry for the confusion. 
I meant: "when **we call e.g.** `self > other` and `other` is dense" not "when `self > other` and `other` is dense".  
You have addressed my concerns completely in  [4924acfb](https://github.com/scipy/scipy/commit/4924acfb36823aea024af231df7b2c56ce2885da).  
Thanks!

perimosocordiae commented:
Now that https://github.com/scipy/scipy/pull/21412 has been merged, we'll need to rebase.

anushkasuyal committed:
repeated method

dschult committed:
Merge branch 'main' into extend-coo-nd-other

dschult committed:
replace __add__ test of broadcasting fail also put in addition part of method

dschult committed:
remove unused test match options and remove whitespace

dschult committed:
Adjust changes that occured during merges

dschult committed:
use isinstance instead of type() ==

dschult committed:
fix idx_dtype error

dschult committed:
replace/fix removed tests. Fix broadcasting in DIA/BSR

dschult commented:
I have merged this with main and resolved conflicts.
I also took the liberty of removing extraneous whitespace at end of lines and simplified some tests for raising exceptions that had options for which string to match in the error message, and I switched code from checking `type(...) == ` to using `isinstance`.  I think all of these are cosmetic changes.

I fixed an index dtype error issue in `_coo.py`. 
I replaced any commented-out tests from this PR. These were mostly `assert_raises` tests that no longer raised due to broadcasting, or only raised on some formats. I expanded handling of broadcasting in BSR and DIA so that they passed. For some commented-out tests I simply uncommented them because ndarray raises and so should sparray. For most, I added a test that the sparse result equals the ndarray result. So for these tests, this PR now replaces tests of broadcasting causing exceptions with tests that the broadcasting case matches ndarray.

I think this is ready to review now/again. I feel like the broadcasting approach is heavy due to copying data to implement broadcasting.  But I think it should be fine to go ahead with this approach now that the 1.15 ship has sailed. We will have time to implement other approaches to broadcasting for 1.16. And if that doesn't work out, we will still be moving ahead because broadcasting is a feature we didn't have before -- and even if it involves copies, that is (probably) better than not having the feature at all.  But I'm hopeful we will some sort of alternative approach in the next few months that avoids copies.

This PR combines adding broadcasting in 2D with adding nD features like `binop`, `min/max` and others. If reviewing is too much we might try splitting it, but let's take a shot. I haven't fully worked through the nD features yet. But I hope to soon.

perimosocordiae commented on scipy/sparse/_coo.py:
```diff
@@ -572,49 +569,62 @@ def eliminate_zeros(self):
-            coords = np.concatenate(self.coords)
-            coo_todense_nd(strides, self.nnz, self.ndim,
-                           coords, self.data, result.ravel('A'), fortran)
+            result_shape = self.shape
+            self = self.reshape(1, -1)
+            other = other.reshape(1, -1)
+            result = _spbase._add_dense(self, other)
+            # reshape back to n-D
+            result = result.reshape(result_shape)
```
This seems like it could be rather expensive, as reshaping a COO array requires reallocating, and the base _add_dense will convert to CSR format and allocate even more.

How about:

```python
self.sum_duplicates()
result[*self.coords] += self.data
```

anushkasuyal replied:
Done! (b619b2b7dc5dfed3daad7a92a5b92d384e54149e)

anushkasuyal replied:
Ah, the unpacking operator is giving syntax error for older python versions I think. Could I try replacing that with `result[tuple(self.coords)] += self.data`? Would that lead to copying of data or does `tuple()` just create an object referencing `self.coords`?

dschult replied:
`self.coords` is already a tuple of 1d arrays. 
So try `result[self.coords] += self.data`. 
That should work for both  Python 3.10 and 3.13

dschult committed:
Merge branch 'main' into extend-coo-nd-other

anushkasuyal committed:
nd _add_dense refactored

anushkasuyal committed:
fixed syntax error

tylerjereddy commented:
I talked to @dschult about this one this morning. My understanding of the "the plan" here is that he and @stefanv will try to review the related work this week, but that it may take the form of two other PRs that have been split off/based on this one.

I believe those two other PRs are not yet milestoned. So, what may happen is that this PR gets closed and superseded by two other separate PRs based on it. That doesn't mean the work done here wasn't important--it was just reorganized a bit in two other PRs for the `sparse` folks to work with.

dschult commented:
This work has been split up to enable easier review. There are two PRs (#22540 and #22987) implementing the nD arithmetic without broadcasting, another PR (#21925) building on the broadcasting approach here. Other features here such as removing some usage of the `sparsetools` routine `coo_todense_nd` will appear in a future PR.

I've removed the milestone from this PR and marked it as draft so it can't easily be merged -- but leaving it open so the branch is sustained until these features are implemented.

Thanks @anushkasuyal Your commits now appear in the other PRs rather than this one, but they are getting merged! :)
